File: modules/agent/src/schemas/report.py
"""Report schemas for API validation."""

from typing import Any, List, Literal, Optional

from pydantic import BaseModel


class ReportFilter(BaseModel):
    field: str
    operator: Literal["eq", "neq", "gt", "gte", "lt", "lte", "contains", "starts_with", "is_null", "is_not_null", "in"]
    value: Any = None


class Aggregation(BaseModel):
    function: Literal["count", "sum", "avg", "min", "max"]
    field: str
    alias: str


class ReportQueryRequest(BaseModel):
    primary_entity: Literal["organizations", "individuals", "contacts", "leads", "notes"]
    columns: List[str]
    joins: Optional[List[str]] = None
    filters: Optional[List[ReportFilter]] = None
    group_by: Optional[List[str]] = None
    aggregations: Optional[List[Aggregation]] = None
    limit: int = 100
    offset: int = 0
    project_id: Optional[int] = None


class SavedReportCreate(BaseModel):
    name: str
    description: Optional[str] = None
    query_definition: dict
    project_ids: Optional[List[int]] = None


class SavedReportUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    query_definition: Optional[dict] = None
    project_ids: Optional[List[int]] = None

File: modules/agent-go/internal/schemas/report.go
package schemas

// ReportFilter represents a filter condition for report queries
type ReportFilter struct {
	Field    string `json:"field" validate:"required"`
	Operator string `json:"operator" validate:"required,oneof=eq neq gt gte lt lte contains starts_with is_null is_not_null in"`
	Value    any    `json:"value"`
}

// Aggregation represents an aggregation function for report queries
type Aggregation struct {
	Function string `json:"function" validate:"required,oneof=count sum avg min max"`
	Field    string `json:"field" validate:"required"`
	Alias    string `json:"alias" validate:"required"`
}

// ReportQueryRequest represents the request body for querying reports
type ReportQueryRequest struct {
	PrimaryEntity string          `json:"primary_entity" validate:"required,oneof=organizations individuals contacts leads notes"`
	Columns       []string        `json:"columns" validate:"required,min=1"`
	Joins         []string        `json:"joins"`
	Filters       []ReportFilter  `json:"filters"`
	GroupBy       []string        `json:"group_by"`
	Aggregations  []Aggregation   `json:"aggregations"`
	Limit         int             `json:"limit" validate:"omitempty,min=1,max=1000"`
	Offset        int             `json:"offset" validate:"omitempty,min=0"`
	ProjectID     *int64          `json:"project_id"`
}

// SavedReportCreate represents the request body for creating a saved report
type SavedReportCreate struct {
	Name            string         `json:"name" validate:"required"`
	Description     *string        `json:"description"`
	QueryDefinition map[string]any `json:"query_definition" validate:"required"`
	ProjectIDs      []int64        `json:"project_ids"`
}

// SavedReportUpdate represents the request body for updating a saved report
type SavedReportUpdate struct {
	Name            *string        `json:"name"`
	Description     *string        `json:"description"`
	QueryDefinition map[string]any `json:"query_definition"`
	ProjectIDs      []int64        `json:"project_ids"`
}
